torch/distributed/_fsdp/fully_sharded_data_parallel.py

```diff
@@ -0,0 +1,803 @@
+                        m._assert_state(TrainingState.BACKWARD_POST)
+                    else:
+                        m._assert_state(TrainingState.BACKWARD_PRE)
+                else:
+                    # When `m` and its children has no params or has params but
+                    # none with `requires_grad==True`, there are two cases:
+                    # 1. output tensors are `requires_grad==True`. In this case,
+                    # pre-backward hook is still registered, so it is in BACKWARD_PRE state.
+                    # 2. output tensors are `requires_grad==False`. In this case,
+                    # pre-backward hook is not registered, so it is in IDLE state.
+                    m._assert_state([TrainingState.BACKWARD_PRE, TrainingState.IDLE])
+                m.training_state = TrainingState.IDLE
+
+                if m._is_root:
+                    # reset this flag for cases like "one forward pass + multiple backward passes"
+                    self._post_backward_callback_queued = False
+
+    @torch.no_grad()
+    def _rebuild_full_params(self) -> None:
+        """
+        Gather all shards of params.
+        """
+        def update_p_data(output_tensor: torch.Tensor) -> None:
+            """
+            Helper function to update p.data pointer.
+            Args:
+                output_tensor (torch.Tensor): this tensor contains the data we just gathered.
+            """
+            p.data = output_tensor
+            # Trim any padding and reshape to match original size.
+            p.data = p.data[: p._orig_size.numel()].view(p._orig_size)  # type: ignore[attr-defined]
+
+        with torch.cuda.stream(self._streams["all_gather"]):
+            for p in self.params:
+                # e.g., when world_size == 1
+                if not p._is_sharded:  # type: ignore[attr-defined]
+                    continue
+                else:
+                    # Move params in CPU to CUDA for the all-gather.
+                    p_data = p.data.to(p._full_param_padded.device, non_blocking=True)  # type: ignore[attr-defined]
+                    p_full_size = p._full_param_padded.size()  # type: ignore[attr-defined]
+                    assert p_full_size.numel() == p_data.numel() * self.world_size
+                    assert p._full_param_padded.storage().size() == 0  # type: ignore[attr-defined]
+                    # Allocate based on full size from all shards.
+                    _alloc_storage(p._full_param_padded, size=p_full_size)  # type: ignore[attr-defined]
+                    output_tensor = p._full_param_padded  # type: ignore[attr-defined]
+
+                    # Fill output_tensor with (p.data for each shard in self.world_size)
+                    dist._all_gather_base(output_tensor, p_data, group=self.process_group)
+
+                    # Set p.data = output_tensor (with padding trimmed)
+                    update_p_data(output_tensor)
```

Comment:
curious, this seems just used once? Any reason for moving it into a dedicated function?

Reply:
I refactored this function a little bit to fix a bug when params are not flatten, forgot remove this function. but I think it is useful to keep it here, it will add more stuff to this function when we add mixed_precision and cpu_offloading